core/lib/utils/src/env.rs

```diff
@@ -0,0 +1,31 @@
+use std::{path::PathBuf, sync::Mutex};
+
+static WORKSPACE: Mutex<Option<PathBuf>> = Mutex::new(None);
+
+pub fn locate_workspace() -> Option<PathBuf> {
+    let mut workspace = WORKSPACE.lock().unwrap_or_else(|x| x.into_inner());
+    if let Some(workspace) = workspace.clone() {
+        return Some(workspace);
+    };
+
+    let output = std::process::Command::new(
+        std::env::var("CARGO")
+            .ok()
+            .unwrap_or_else(|| "cargo".to_string()),
+    )
+    .arg("locate-project")
+    .arg("--workspace")
+    .output()
+    .ok()?;
```

Comment:
I'm not sure I agree with swallowing errors here; it means that if something is misconfigured, we'd repeatedly run into the same errors. Note that `insta` turns errors into panics, which is the reasoning behind unwrapping mutex poison errors.

Reply:
so, it's likely won't work with pure binary and that's why we have a fallback

Reply:
Ah, I see. It'd still make sense to log errors, IMO.

Reply:
I can log it on a trace level, because they doesn't make sense in production